extensions/rvadapter/src/main/java/com/google/android/agera/rvadapter/RepositoryPresenterCompiler.java

```diff
@@ -101,165 +110,57 @@ public Object recycleWith(@NonNull final Receiver recycler) {
     return this;
   }
 
-  private abstract static class BasicRepositoryPresenter<TVal, T>
-      extends RepositoryPresenter<T> {
+  private static final class CompiledRepositoryPresenter extends RepositoryPresenter {
+    @NonNull
+    private final Function<Object, List<Object>> converter;
     @NonNull
     private final Function<Object, Integer> layoutId;
     @NonNull
-    private final Binder<TVal, View> binder;
+    private final Binder<Object, View> binder;
     @NonNull
-    private final Function<TVal, Long> stableIdForItem;
+    private final Function<Object, Long> stableIdForItem;
     @NonNull
     private final Receiver<View> recycler;
+    private List items = emptyList();
 
-    BasicRepositoryPresenter(@NonNull final Function<Object, Integer> layoutId,
-        @NonNull final Binder<TVal, View> binder,
+    CompiledRepositoryPresenter(
+        @NonNull final Function<Object, Integer> layoutId,
+        @NonNull final Binder<Object, View> binder,
+        @NonNull final Function<Object, Long> stableIdForItem,
         @NonNull final Receiver<View> recycler,
-        @NonNull final Function<TVal, Long> stableIdForItem) {
-      this.layoutId = checkNotNull(layoutId);
-      this.binder = checkNotNull(binder);
-      this.recycler = checkNotNull(recycler);
+        @NonNull final Function<Object, List<Object>> converter) {
+      this.converter = converter;
+      this.layoutId = layoutId;
+      this.binder = binder;
       this.stableIdForItem = stableIdForItem;
+      this.recycler = recycler;
     }
 
-    @NonNull
-    protected abstract TVal getValue(@NonNull final T data, final int index);
-
     @Override
-    public final int getLayoutResId(@NonNull final T data, final int index) {
-      return layoutId.apply(getValue(data, index));
+    public int getItemCount(@NonNull final Object data) {
+      items = converter.apply(data);
```

Comment:
I'm thinking this may need to be replicated everywhere -- with a cache of the `data` object (maybe using a WeakReference) and hidden within a private method getItems(data). Just for robustness.